### DISCUSSION TITLE
Prisma's typings break after setting `strictNullChecks` to `true`

### DISCUSSION THREAD

**Discussion by aprilmintacpineda** [prisma/prisma#21782] [Q&A] [answered]
### Question

I have the following function somewhere:

```ts
getAll(
  skip: number,
  options?: {
    include?: Prisma.VacancyFindManyArgs['include'];
    count?: boolean;
  }
) {
  const { count, include } = options || {};

  if (count) {
    return this.prismaService.$transaction([
      this.prismaService.vacancy.count(),
      this.prismaService.vacancy.findMany({
        skip,
        take: 100,
        include
      })
    ]);
  }

  return this.prismaService.vacancy.findMany({
    skip,
    take: 100,
    include
  });
}
```

And the following code, somwhere else, that calls that function

```ts
const result = await getAll(skip, {
  count: true,
  include: {
    applicationForm: true,
    company: true
  }
});

if (!result) throw new InternalServerErrorException();

const [count, vacancies] = result;

return vacancies.map((_vacancy) => {
  const vacancy = doSomethingWithVacancy(_vacancy);

  return {
    ...vacancy,
    company: doSomethingWithCompany(vacancy.company),
  };
});
```

Everything works well as expected, I get the proper typings and no issues. However, I just realized that NestJS initially sets `strictNullChecks` to `false` which I dislike because what's the point of using typescript if that's the case. You'd given up a strong feature of typescript that makes you write less buggy code. So I set it to `true` and then I started getting:

```
Property 'map' does not exist on type '({ id: string; title: string; description: string; salaryRate: number; salaryFrequency: string; salaryCurrency: string; workSetup: string; contractType: string; timezone: string; equipmentSetup: string; ... 10 more ...; deletedAt: Date | null; } & { ...; }) | ({ ...; } & { ...; })[]'.
  Property 'map' does not exist on type '{ id: string; title: string; description: string; salaryRate: number; salaryFrequency: string; salaryCurrency: string; workSetup: string; contractType: string; timezone: string; equipmentSetup: string; ... 10 more ...; deletedAt: Date | null; } & { ...; }'.ts(2339)
```

I also noticed that the types for `include` basically just disappears because I get an error somewhere else and it's because of this code:

```ts
this.userRepository.findById(userId, {
  include: {
    company: true
  }
})
```

Notice how I have `company: true` which means I should get `User & { company: Company }` -- which I was prior to setting `strictNullChecks` to `true`, but afterwards, that won't be the case anymore, I would only get `User`, so doing:

```ts
const user = await this.userRepository.findById(userId, {
  include: {
    company: true
  }
});

user.company;
```

Would yield the following error:

```
Property 'company' does not exist on type '{ id: string; email: string; hashedPassword: string; firstName: string | null; middleName: string | null; lastName: string | null; lastActiveAt: Date | null; emailVerifiedAt: Date | null; ... 5 more ...; deletedAt: Date | null; }'. Did you mean 'companyId'?
```

### How to reproduce (optional)

N/A

### Expected behavior (optional)

Typings shouldn't break when `strictNullChecks` is set to `true`

EDIT:

Turns out the issue with the `getAll` was a problem with my code, for some reason, the proper type is not being inferred based on `count`. But the problem with `include` is indeed with because of setting `strictNullChecks` to `true`

EDIT:

I think the problem is with the function itself, somehow the type inference gets lost with `strictNullChecks: true`.

```ts
findById (
  id: string,
  options?: {
    include?: Prisma.UserFindUniqueArgs['include'];
  }
) {
  const { include } = options || {};

  return this.prismaService.user.findUnique({
    where: { id },
    include
  });
}

// then somewhere else

const user = await findById(id);
user.company // <--- results to an error
```

But if I call prisma directly:

```ts
const user = await this.prismaService.user.findUnique({
  where: { id },
  include: {
    company: true
  }
});

user?.company // <--- no error
```

**Comment by ludralph** [comment, marked as answer]
Hi @aprilmintacpineda 👋 

Thank you for raising this question.

> Turns out the issue with the getAll was a problem with my code, for some reason, the proper type is not being inferred based on count. But the problem with include is indeed with because of setting strictNullChecks to true.

When `strictNullChecks` is enabled, TypeScript becomes more rigorous in its handling of null and undefined values, which can sometimes lead to type inference issues.

In the case of the `getAll` function, the problem is due to the conditional return statement. When `count` is true, the function returns the result of a transaction, which is a tuple containing the count and the vacancies. However, when `count` is false, the function returns the vacancies directly. This difference in the return type can confuse TypeScript's type inference, leading to the error you see. You can explicitly specify the return type of the getAll function. This will provide TypeScript with more information about the possible return values and help it infer the correct type for the result variable. Here's an example:

```ts
async getAll(
  skip: number,
  options?: {
    include?: Prisma.VacancyFindManyArgs['include'];
    count?: boolean;
  }
): Promise<(number & Prisma.Vacancy[]) | Prisma.Vacancy[]> {
  const { count, include } = options || {};

  if (count) {
    return this.prismaService.$transaction([
      this.prismaService.vacancy.count(),
      this.prismaService.vacancy.findMany({
        skip,
        take: 100,
        include
      })
    ]);
  }

  return this.prismaService.vacancy.findMany({
    skip,
    take: 100,
    include
  });
}
```
With this explicit return type, TypeScript should be able to correctly infer the type of the result variable and avoid the error.



> I think the problem is with the function itself, somehow the type inference gets lost with strictNullChecks: true.

With regards the `findById` function, the issue is similar. When `strictNullChecks` is enabled, TypeScript is more cautious about accessing properties on objects that may be null or undefined. In the case of the user object returned by the findById function, its company property may be null if the user doesn't have an associated company.

To handle this, you can either use the optional chaining operator (?.) to safely access the company property, or you can explicitly check for null before accessing it. 

```ts
async findById(
  id: string,
  options?: {
    include?: Prisma.UserFindUniqueArgs['include'];
  }
) {
  const { include } = options || {};

  return this.prismaService.user.findUnique({
    where: { id },
    include
  });
}

// Using optional chaining
const user = await findById(id);
const company = user?.company;

// Checking for null
const user = await findById(id);
if (user) {
  const company = user.company;
} else {
  // Handle the case where user is null
}
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
